### DISCUSSION TITLE
cant download linux agent

### DISCUSSION THREAD

**Discussion by HozerTech** [amidaware/tacticalrmm#1112] [Q&A] [answered]
when i try to download the install script for a linux agent i get an error 'Missing Code Signing Token 400:' 
it doesn't matter what selections i make for the linux agent. 

I can download the installer for windows just fine though. 

any suggestions on getting the linux agent to download?

**Comment by wh1te909** [comment, marked as answer]
Hi, https://docs.tacticalrmm.com/faq/#linux-agent and https://docs.tacticalrmm.com/code_signing/

**Reply by HozerTech** [reply]
Thank you, i get that. guess i was just hoping to have a way to test a linux agent prior to committing to the $50/month. 
in the end, if it works well enough and is stable id deploy to production and the code signing is a great thing to have.

**Reply by wh1te909** [reply]
you can compile the linux agent and deploy it yourself for testing, the code is all in the agent repo. all code signing does it just make it easy with 1 click

**Reply by HozerTech** [reply]
ah, guess i missed that. thank you.

**Reply by BelleNottelling** [reply]
Be sure to use / reference this script so you properly install it
https://github.com/amidaware/tacticalrmm/blob/develop/api/tacticalrmm/core/agent_linux.sh

I deployed this on a good handful of my Linux boxes without issue.
It's handy to ensure they all stay up without an issue

**Reply by HozerTech** [reply]
thank you. 
now when i try to build the agent i get a checksum mismatch.

**Comment by HozerTech** [comment]
got the agent to build, go apparently didnt like my fedora system...?? stood up an ubuntu vm and installed go and it built perfectly despite the same version (1.18.1) as on my fedora box.

**Reply by dinger1986** [reply]
Most likely a go issue, can raise a bug with them.

Did you try and install the agent?

**Reply by HozerTech** [reply]
I did try to install the agent, it seems to be getting hung up on downloading/installing the mesh agent - 
i have meshDL='https://mesh.mydomain.com/' and i get this out put - 
`Downloading tactical agent...
Downloading and installing mesh agent...
/tmp/mesh-PHn611m26/meshagent: line 1: !DOCTYPE: No such file or directory
/tmp/mesh-PHn611m26/meshagent: line 2: html: No such file or directory
/tmp/mesh-PHn611m26/meshagent: line 3: head: No such file or directory
/tmp/mesh-PHn611m26/meshagent: line 4: meta: No such file or directory
/tmp/mesh-PHn611m26/meshagent: line 5: meta: No such file or directory
/tmp/mesh-PHn611m26/meshagent: line 6: meta: No such file or directory
/tmp/mesh-PHn611m26/meshagent: line 7: meta: No such file or directory
/tmp/mesh-PHn611m26/meshagent: line 8: meta: No such file or directory
/tmp/mesh-PHn611m26/meshagent: line 9: meta: No such file or directory
/tmp/mesh-PHn611m26/meshagent: line 10: link: No such file or directory
/tmp/mesh-PHn611m26/meshagent: line 11: link: No such file or directory
/tmp/mesh-PHn611m26/meshagent: line 12: link: No such file or directory
/tmp/mesh-PHn611m26/meshagent: line 13: syntax error near unexpected token `<'
'tmp/mesh-PHn611m26/meshagent: line 13: `    <script type="text/javascript" src="scripts/common-0.0.1.js"></script>
Getting mesh node id...
Created symlink /etc/systemd/system/multi-user.target.wants/tacticalagent.service → /etc/systemd/system/tacticalagent.service.
`

is the mesh agent not located at mesh.mydomain.com that was setup durring the RMM server install?

**Reply by dinger1986** [reply]
yes but you need to get the url for downloading not put in the main url

**Reply by HozerTech** [reply]
ok, i tried meshDL='https://mesh.mydomain.com/meshagents?script=1'

and get
`Downloading tactical agent...
Downloading and installing mesh agent...
Device group identifier is not correct, must be at least 64 characters long.
Getting mesh node id...
Created symlink /etc/systemd/system/multi-user.target.wants/tacticalagent.service → /etc/systemd/system/tacticalagent.service.
`

**Reply by HozerTech** [reply]
if i remove the ?script=1 it seems to finish but nothing shows in the rmm
'Downloading tactical agent...
Downloading and installing mesh agent...
Getting mesh node id...
Created symlink /etc/systemd/system/multi-user.target.wants/tacticalagent.service → /etc/systemd/system/tacticalagent.service.
'

**Reply by dinger1986** [reply]
log into your mesh web gui, click add agent, then select Linux / BSD / macOS Binary, select your other preferred options then copy the command at the bottom, you want to put in everything between the ". I literally just figured that out by going to mesh myself so should be simple enough.

**Reply by HozerTech** [reply]
I thought I had tried that but I must have done the Linux/BSD installer and not the binary. 
I think it installed ok.
`Downloading tactical agent...
Downloading and installing mesh agent...
...Checking for previous installation of "meshagent" [NONE]
...Installing service [DONE]
   -> Starting service... [OK]
Getting mesh node id...
Created symlink /etc/systemd/system/multi-user.target.wants/tacticalagent.service → /etc/systemd/system/tacticalagent.service.
`
but i still dont see it in RMM...

This is what i have all the variables in the script set to.
```
agentDL='https://mydomain.com/rmmagent'
meshDL='https://mesh.mydomain.com/meshagents?id=uMp2PbXOrM5szHFdLEewMoAFeuNbn3BAHniMORYnGm4Xc2ASxFNVKoKZ%24NRLy1%24o&installflags=0&meshinstall=6'

apiURL='https://api.mydomain.com'
token='efa0ef353d4c654d444427702fa65ba9d6a6628888bdb812b8aabc123e395834f75'
clientID='1'
siteID='1'
agentType='workstation'
proxy=''
```
the only other thing would maybe be the token? where should i be getting that?
I had originally downloaded the windows powershell script from my server and copied it out of there.

I have other issues with Meshcentral but i think they are all proxy related at the moment.

**Reply by dinger1986** [reply]
Maybe want to delete your token? Unless it's expired after 24 hours?

Is the trmm and mesh agents downloading?

**Reply by HozerTech** [reply]
they are downloading and installing. I just commented out the token line and ran the installer again. it removed the mesh agent and reinstalled it. I also checked status or tacticalagent.service and its failing to start.
this is all i can find in the logs 

>tacticalagent.service: Failed at step EXEC spawning /usr/local/bin/>
>tacticalagent.service: Main process exited, code=exited, status=203/EXEC
>tacticalagent.service: Failed with result 'exit-code'.

**Reply by dinger1986** [reply]
Try and get a fresh token and try it as that ones maybe expired if you only set it to 24 hours.

**Reply by HozerTech** [reply]
same issue. 
windows agents do install and connect just fine. so its either my install script or agent install didnt compile properly

**Reply by dinger1986** [reply]
Is your install script creating a service? Does mesh agent now show up?

**Reply by HozerTech** [reply]
yes, i see meshagent service running but tacticalagent service is not starting. 
i do not see the system under devices in the mesh interface or in TRMM interface

**Reply by dinger1986** [reply]
I don't know, you mentioned proxys as well so dunno if that's causing issues.

What install script are you using for the linux agent?

**Comment by HozerTech** [comment]
this is the install script i'm using - 
https://github.com/amidaware/tacticalrmm/blob/develop/api/tacticalrmm/core/agent_linux.sh

the end of the day if i can get mesh working, when i try remote desktop the screen is blank and remote terminal and file browser are also blank. 
once i get those bugs worked out I'll likely just pay the $50 a month for the signing token since TRMM is really better than any paid RMM i have used an honestly has way better support :)

**Reply by dinger1986** [reply]
May I suggest joining discord, it's a very active community and plenty of users who can help with proxy's etc

**Comment by HozerTech** [comment]
I'll likely do that. Thank you for all your help.


________________________________

**Comment by Erinomos** [comment]
Hello, could you give an example of the parts to modify in the script?

**Reply by dinger1986** [reply]
as above, theres guide and plenty of people helped on discord, theres afew threads as well

**Comment by Erinomos** [comment]
Could you put the link of the guide because I do not speak English. It will be difficult for me to understand the discussions on discord.

**Reply by dinger1986** [reply]
there is no guide just what others have done on discord, I will ask for you if someone can come and help, what language do you speak?

**Comment by Erinomos** [comment]
Je parle Français.

**Reply by dinger1986** [reply]
ok shall ask

**Reply by Scaff31** [reply]
> Je parle Français.

Salut. Peux tu m'expliquer ton problème et le résultat que tu souhaites obtenir.

Je t'aiderai avec plaisir.

**Comment by Erinomos** [comment]
Bonjour Scaff, je cherche quel information mettre dans le script d'installation de l'agent linux:
agentDL='https://rmm.exemple.fr/'
meshDL='https://mesh.exemple.fr/meshagents?id=fF7oV21DzGCSJbkqKzvMW%400v>

apiURL='https://api.exemple.fr'
token='ed0faf7128c324949660db0baa669ce9e5c3789637f2b7d1c98df1c6e9d983d8'
clientID='1'
siteID='1'
agentType='server'
proxy=''
J'ai du  loupé quelque chose.

**Reply by Scaff31** [reply]
Ok donc voici les infos qu'il te faut et/ou comment les obtenir.

Agentdl= un lien de téléchargement du binaire linux
Meshdl= un lien de téléchargement du binaire meshcentral

Apiurl= api.tondomain.com

Le reste des informations tu peux les obtenir depuis l'interface tactical rmm. Tu va dans agent -> install-> tu choisi tes options (groupe, type, etc ...) Puis en bas tu sélectionne installation manuel via powershell de la tu pourra récupérer le token, l'id du site, id client, le type

**Reply by Erinomos** [reply]
pour Agentdl il faut avoir télécharger le binaire sur github? car je ne trouve que l'agent windows sur github.

**Reply by Scaff31** [reply]
Si tu ne fait pas partie des sponsor il fait que tu le compile toi même.

**Reply by Erinomos** [reply]
ok, si je deviens sponsor, je pourrais l'installer directement à partir de l'interface rmm comme pour l'agent windows?

**Reply by Scaff31** [reply]
Oui tout a fait il est compilé par l'équipe qui développe tactical. A savoir qu'il sera disponible gratuitement pour l'ensemble des utilisateurs une fois qu'il ne sera plus en beta

**Reply by Scaff31** [reply]
Je crois que des utilisateurs on compilé l'agent et qu'il le rende disponible sur github mais je n'ai plus le nom il faut demander sur le discord

**Reply by Erinomos** [reply]
Est ce que tu as une idée de la date de fin de la beta? car j'ai juste besoin de le tester sur une machine pour le moment.

**Reply by Scaff31** [reply]
Non je ne sais pas du tout après la compilation de l'agent c'est très simple c'est une commande à faire sous linux

**Reply by Erinomos** [reply]
Ok, je vais voir ce que je peux faire. Merci pour ces informations et bonne journée.

**Reply by Scaff31** [reply]
Je peux te compiler l'agent mais pas avant demain

**Reply by Erinomos** [reply]
pas de soucis, merci.

**Reply by Serenfo** [reply]
@Erinomos Peux tu me dire qu'elle architecture tu veux pour l'agent ? sur qu'elle type de machine tu veux l'installer ?

**Reply by Serenfo** [reply]
Building and installing linux agent TRMM

1 - Build
    - Install golang on your machine 
        https://go.dev/doc/install
    - Get the github of the agent
        git clone https://github.com/amidaware/rmmagent
       or for Windows: from the agent repository click on "Code" -> "Download ZIP".
    - Go to the rmmagent folder
        env CGO_ENABLED=0 GOOS=<GOOS> GOARCH=<GOARCH> go build -ldflags "-s -w"
        https://gist.github.com/asukakenji/f15ba7e588ac42795f421b48b8aede63

2 - Host the rmmagent generator file it must be downloadable from internet
3 - Modify the installation script
    https://github.com/amidaware/tacticalrmm/blob/develop/api/tacticalrmm/core/agent_linux.sh

    agentDL='agentDLChange' # Paste the link of your rmmagent
    meshDL='meshDLChange' # Paste the link provided from the meshcentral web interface ( mesh.yourdomain.com ) from a group -> "Add an agent" -> Linux/BSD binaries

    apiURL='apiURLChange' # https://api.yourdomain.com/
    token='tokenChange' # Get a token from TRMM interface -> Install Agent -> Windows and select manual -> -auth=token
    clientID='clientIDChange' # Get the TRMM interface -> Install Agent -> Windows and select manual -> -client-id=
    siteID='siteIDChange' # Get TRMM interface -> Install Agent -> Windows and select manual -> -site-id=
    agentType='agentTypeChange' # server or workstation
    proxy='''

**Reply by Erinomos** [reply]
Bonjour, j'arrive bien à compiler l'agent (avec l'aide de scaff), l'installation manuelle de l'agent fonctionne mais pas avec le script/
je pense que le problème viens de:
agentDL: '/usr/locale/rmmagent'
meshDL='/usr/locale/meshagent'

je met le binaire de l'agent et meshagent directement sur la machine hote.

**Reply by Serenfo** [reply]
agentDL='agentDLChange' # Paste the link of your rmmagent
meshDL='meshDLChange' # Paste the link provided from the meshcentral web interface ( mesh.yourdomain.com ) from a group -> "Add an agent" -> Linux/BSD binaries

AgentDL = ça doit être un lien sur internet d'ou le script peux télécharger le binaire ( github, cloud, etc... )
meshDL='meshDLChange' # Copie le lien fourni par l'interface web de meshcentral ( mesh.tondomain.com ) depuis un groupe => "Ajouter un agent" => tu choisi linux/BSD

**Reply by Erinomos** [reply]
Merci, je test dès que possible.
